p2p/state.py

```diff
@@ -101,74 +101,95 @@ def register_peer(self, peer: BasePeer) -> None:
             self.logger.debug("Waiting for %d running peers to finish", len(self._running_peers))
             await asyncio.sleep(0.1)
 
-    async def request_next_batch(self):
-        requests = self.scheduler.next_batch(eth.MAX_STATE_FETCH)
-        if not requests:
-            # Although our run() loop frequently yields control to let our msg handler process
-            # received nodes (scheduling new requests), there may be cases when the pending nodes
-            # take a while to arrive thus causing the scheduler to run out of new requests for a
-            # while.
-            self.logger.debug("Scheduler queue is empty, not requesting any nodes")
-            return
-        self.logger.debug("Requesting %d trie nodes", len(requests))
-        await self.request_nodes([request.node_key for request in requests])
-
     async def request_nodes(self, node_keys: List[bytes]) -> None:
         # FIXME: Need a better criteria to select peers here.
-        peer = await self.peer_pool.get_random_peer()
+        peer = await wait_with_token(self.peer_pool.get_random_peer(), token=self.cancel_token)
         now = time.time()
         for node_key in node_keys:
             self._pending_nodes[node_key] = now
+        self.logger.debug("Requesting %d trie nodes", len(node_keys))
         cast(ETHPeer, peer).sub_proto.send_get_node_data(node_keys)
 
-    async def retry_timedout(self):
-        timed_out = []
-        now = time.time()
-        for node_key, req_time in list(self._pending_nodes.items()):
-            if now - req_time > self._reply_timeout:
-                timed_out.append(node_key)
-        if not timed_out:
-            return
-        self.logger.debug("Re-requesting %d trie nodes", len(timed_out))
-        await self.request_nodes(timed_out)
+    async def _periodically_retry_timedout(self):
+        while True:
+            now = time.time()
+            oldest_request_time = now
+            timed_out = []
+            for node_key, req_time in list(self._pending_nodes.items()):
+                if now - req_time > self._reply_timeout:
+                    timed_out.append(node_key)
+                elif req_time < oldest_request_time:
+                    oldest_request_time = req_time
+            if timed_out:
+                self.logger.debug("Re-requesting %d trie nodes", len(timed_out))
+                self._total_timeouts += len(timed_out)
+                await self.request_nodes(timed_out)
+
+            now = time.time()
+            sleep_until = (oldest_request_time + self._reply_timeout) - now
```

Comment:
I think `sleep_for` or `sleep_seconds` or `sleep_duration` would be a more appropriate name as I *think* this is an offset, rather than a timestamp.

Reply:
Indeed, that's an offset. Went with `sleep_duration`